File: hibernate-query-language/src/main/java/io/github/robertomike/hefesto/utils/ReflectionUtils.java
package io.github.robertomike.hefesto.utils;

import lombok.NoArgsConstructor;

import java.beans.IntrospectionException;
import java.beans.Introspector;
import java.beans.PropertyDescriptor;
import java.lang.reflect.AccessibleObject;
import java.lang.reflect.Method;

@NoArgsConstructor(access = lombok.AccessLevel.PRIVATE)
public final class ReflectionUtils {
    public static void makePublic(AccessibleObject accessibleObject) {
        if (accessibleObject != null) {
            accessibleObject.setAccessible(true);
        }
    }

    /**
     * Try to find a class getter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a getter
     * @param propertyName a property name
     * @return the getter method or null, if such getter is not exist
     */
    public static Method getClassGetter(Class<?> classToCheck, String propertyName) {
        PropertyDescriptor[] descriptors = getPropertyDescriptors(classToCheck);

        for (PropertyDescriptor descriptor : descriptors) {
            if (isGetter(descriptor, propertyName)) {
                return descriptor.getReadMethod();
            }
        }

        return null;
    }

    private static boolean isGetter(PropertyDescriptor descriptor, String propertyName) {
        Method method = descriptor.getReadMethod();
        return method != null && method.getParameterTypes().length == 0
                && descriptor.getName().equalsIgnoreCase(propertyName);
    }

    /**
     * Try to find a class setter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a setter
     * @param propertyName a property name
     * @param getterMethod a getter method for getting a type of a property
     * @return the setter method or null, if such setter is not exist
     */
    public static Method getClassSetter(Class<?> classToCheck, String propertyName,
                                        Method getterMethod) {
        return getClassSetter(classToCheck, propertyName,
                getterMethod == null ? null : getterMethod.getReturnType());
    }

    /**
     * Try to find a class setter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a setter
     * @param propertyName a property name
     * @param propertyType a type of a property
     * @return the setter method or null, if such setter is not exist
     */
    public static Method getClassSetter(Class<?> classToCheck, String propertyName,
                                        Class<?> propertyType) {
        PropertyDescriptor[] descriptors = getPropertyDescriptors(classToCheck);

        for (PropertyDescriptor descriptor : descriptors) {
            if (isSetter(descriptor, propertyName, propertyType)) {
                return descriptor.getWriteMethod();
            }

        }

        return null;
    }

    private static boolean isSetter(PropertyDescriptor descriptor, String propertyName,
                                    Class<?> propertyType) {
        Method method = descriptor.getWriteMethod();

        return method != null && method.getParameterTypes().length == 1
                && method.getName().startsWith("set")
                && descriptor.getName().equalsIgnoreCase(propertyName)
                && (propertyType == null || method.getParameterTypes()[0].equals(propertyType));
    }

    private static PropertyDescriptor[] getPropertyDescriptors(Class<?> beanClass) {
        try {
            return Introspector.getBeanInfo(beanClass).getPropertyDescriptors();
        } catch (IntrospectionException ex) {
            throw new RuntimeException(ex);
        }
    }

    public static String[] getPropertyParts(String property) {
        return StringUtils.splitByDot(property);
    }
}

File: hibernate-query-language/src/main/kotlin/io/github/robertomike/hefesto/utils/ReflectionUtils.kt
package io.github.robertomike.hefesto.utils

import java.beans.Introspector
import java.beans.PropertyDescriptor
import java.lang.reflect.AccessibleObject
import java.lang.reflect.Method

object ReflectionUtils {
    @JvmStatic
    fun makePublic(accessibleObject: AccessibleObject?) {
        accessibleObject?.isAccessible = true
    }

    /**
     * Try to find a class getter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a getter
     * @param propertyName a property name
     * @return the getter method or null, if such getter is not exist
     */
    @JvmStatic
    fun getClassGetter(classToCheck: Class<*>, propertyName: String): Method? {
        val descriptors = getPropertyDescriptors(classToCheck)

        for (descriptor in descriptors) {
            if (isGetter(descriptor, propertyName)) {
                return descriptor.readMethod
            }
        }

        return null
    }

    private fun isGetter(descriptor: PropertyDescriptor, propertyName: String): Boolean {
        val method = descriptor.readMethod
        return method != null && method.parameterTypes.isEmpty()
                && descriptor.name.equals(propertyName, ignoreCase = true)
    }

    /**
     * Try to find a class setter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a setter
     * @param propertyName a property name
     * @param getterMethod a getter method for getting a type of a property
     * @return the setter method or null, if such setter is not exist
     */
    @JvmStatic
    fun getClassSetter(
        classToCheck: Class<*>,
        propertyName: String,
        getterMethod: Method?
    ): Method? {
        return getClassSetter(
            classToCheck, propertyName,
            getterMethod?.returnType
        )
    }

    /**
     * Try to find a class setter method by a property name. Don't check parent classes or
     * interfaces.
     *
     * @param classToCheck a class in which find a setter
     * @param propertyName a property name
     * @param propertyType a type of a property
     * @return the setter method or null, if such setter is not exist
     */
    @JvmStatic
    fun getClassSetter(
        classToCheck: Class<*>,
        propertyName: String,
        propertyType: Class<*>?
    ): Method? {
        val descriptors = getPropertyDescriptors(classToCheck)

        for (descriptor in descriptors) {
            if (isSetter(descriptor, propertyName, propertyType)) {
                return descriptor.writeMethod
            }
        }

        return null
    }

    private fun isSetter(
        descriptor: PropertyDescriptor,
        propertyName: String,
        propertyType: Class<*>?
    ): Boolean {
        val method = descriptor.writeMethod

        return method != null && method.parameterTypes.size == 1
                && method.name.startsWith("set")
                && descriptor.name.equals(propertyName, ignoreCase = true)
                && (propertyType == null || method.parameterTypes[0] == propertyType)
    }

    private fun getPropertyDescriptors(beanClass: Class<*>): Array<PropertyDescriptor> {
        return try {
            Introspector.getBeanInfo(beanClass).propertyDescriptors
        } catch (ex: Exception) {
            throw RuntimeException(ex)
        }
    }

    @JvmStatic
    fun getPropertyParts(property: String): Array<String> {
        return StringUtils.splitByDot(property)
    }
}
